fix: flags & environmental variables added for the CLI to configure pod selector labels (#10200)

maheshbaliga opened the pull request:
Fixes #10200 
ArgoCD component names were hard coded as part of the pod selector labels. These values have been made configurable in the CLI by adding new flags and environmental variables.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

keskad commented:
This PR may help ArgoCD operator users who are using multiple ArgoCD instances.

keskad commented:
I believe you should revert the reformatting of the whole repository as it would cause a lot of merge conflicts.

crenshaw-dev commented:
I'm a little concerned about adding a global flag, since it'll be added to the docs for commands that won't use the flag.

But otoh, adding the flags individually to each relevant command is tedious. And it also means the CLI will throw an error when the flag is present but not supported, meaning CI scripts can't simply add the flags to every command to be safe.

Maybe we should just support an env var rather than a flag?

keskad commented:
Environment variable could be enough for me.

maheshbaliga commented:
Hi @crenshaw-dev, How do I proceed with this code fix? 
- Should I only add it to the commands that would need it? I can try, to the best of my knowledge.
- Or Should I remove the flags totally?

crenshaw-dev commented:
@maheshbaliga I'm fine either way. Personally I think a well-documented env var is sufficient. But some folks might prefer flags. Whatever you like better. :-)

maheshbaliga commented:
@crenshaw-dev I've removed the global flags and added flags to the individual commands. Kindly re-review and let me know if anything else has to be added.

crenshaw-dev reviewed (changes requested):
This is looking really really good.

Two (hopefully small) requests:

1) Can you centralize the definition of the doc strings? If we ever need to change, for example, the doc string associated with the application-controller-name param, we'd have to change it in N places.
2) Could you add a docs page to describe what these params do, and when they're intended to be used? We could link to that docs page from the error message in portforwarder.go.

maheshbaliga commented:
@crenshaw-dev Constants have been added for the environmental variable and component names. A link to the [Environmental Variables](https://argo-cd.readthedocs.io/en/stable/user-guide/environment-variables/) page had also been added in the portforwarder.go error message.

RyanW8 commented:
Could you avoid adding flags/env vars entirely? We could instead just pass in multiple `podSelectors` but instead select based on `app.kubernetes.io/component` and `app.kubernetes.io/part-of`.  This should also work when users have multiple ArgoCD installs in the same cluster. As long as users kube context is set to the right namespace/cluster

```
▶ helm install test argo/argo-cd --set nameOverride=toto
<<output trimmed>>
▶ kgp test-toto-repo-server-6655c56bd5-g8czp -oyaml | yq .metadata.labels
app.kubernetes.io/component: repo-server
app.kubernetes.io/instance: test
app.kubernetes.io/managed-by: Helm
app.kubernetes.io/name: toto-repo-server
app.kubernetes.io/part-of: argocd
helm.sh/chart: argo-cd-5.20.0
pod-template-hash: 6655c56bd5
▶ kgp test-toto-redis-5497d95dbd-tfqv4 -oyaml | yq .metadata.labels
app.kubernetes.io/component: redis
app.kubernetes.io/instance: test
app.kubernetes.io/managed-by: Helm
app.kubernetes.io/name: toto-redis
app.kubernetes.io/part-of: argocd
helm.sh/chart: argo-cd-5.20.0
pod-template-hash: 5497d95dbd                                                                                                                                                                                                                                                                   
▶ kgp test-toto-application-controller-0 -oyaml | yq .metadata.labels
app.kubernetes.io/component: application-controller
app.kubernetes.io/instance: test
app.kubernetes.io/managed-by: Helm
app.kubernetes.io/name: toto-application-controller
app.kubernetes.io/part-of: argocd
controller-revision-hash: test-toto-application-controller-6bb5bfd8c5
helm.sh/chart: argo-cd-5.20.0
statefulset.kubernetes.io/pod-name: test-toto-application-controller-0
```

Diliz commented:
any news regarding this one? @maheshbaliga

maheshbaliga commented:
Hi @Diliz, I've implemented the changes as required by @crenshaw-dev. Waiting for his re-review/approval.

Diliz commented:
up

Diliz commented:
Hello!

@crenshaw-dev @maheshbaliga

Is it mergeable or something else is needed here?

@RyanW8  
Does it work for the case you mentionned? (I think it should as it is, but maybe I got something wrong)

maheshbaliga commented:
@Diliz, Michael wants me to revert the changes back to the global flags implementation. Refer: [slack chat](https://cloud-native.slack.com/archives/C020XM04CUW/p1669898808924529). I just need time to revert the changes and implement it again because I had not saved the code changes made for the global flags.

keskad commented:
@maheshbaliga Thanks for your good work :)

jacksgt commented:
Thanks for the excellent work. We are also really looking forward to this fix. :-)
An env var would be totally sufficient for us.

crenshaw-dev commented:
I've attempted to re-implement this with a global flag here: https://github.com/argoproj/argo-cd/pull/14605

andrii-korotkov-verkada commented:
Closed the PR due to it being old and unlikely to be able to resolve merge conflicts. But feel free to re-open/re-submit if the work is still planned.
